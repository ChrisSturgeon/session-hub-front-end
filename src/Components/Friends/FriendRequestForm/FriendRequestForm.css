.add-friend {
  display: flex;
  width: 180px;
}

.add-friend button {
  width: 90px;
}

.add-btn {
  width: 180px;
}

.add-btn,
.add-friend .confirm-btn {
  background-color: var(--dark-blue);
  color: white;
  font-weight: 500;
  border: none;
  outline: none;
  padding: 0.3em;
  border-radius: 4px;
}

.add-friend .confirm-btn {
  width: 90px;
}

.add-friend .cancel-btn {
  width: 180px;
  background-color: #01182611;
  color: black;
  font-weight: 500;
  border: none;
  outline: none;
  padding: 0.3em;
  border-radius: 4px;
}
