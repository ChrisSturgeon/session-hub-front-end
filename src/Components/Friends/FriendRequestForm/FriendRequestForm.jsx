import './FriendRequestForm.css';
import { useState } from 'react';

export default function FriendRequestForm({ userID }) {
  const [showForm, setShowForm] = useState(false);
  const [requestSent, setRequestSent] = useState(null);
  const [addButtonText, setAddButtonText] = useState('Add Friend');

  const flip = (event) => {
    if (event) {
      event.preventDefault();
    }
    setShowForm(!showForm);
  };

  // Flips button to show confirmation prompt
  if (!showForm) {
    return (
      <button
        className="add-btn"
        onClick={(event) => {
          flip(event);
        }}
        disabled={requestSent}
      >
        {addButtonText}
      </button>
    );
  }

  // Sends POST request to API route
  const handleSubmit = async (event) => {
    event.preventDefault();
    console.log('Submitting friend request...');
    const url = 'http://localhost:3000/api/friends/request/create';

    try {
      const response = await fetch(url, {
        method: 'POST',
        mode: 'cors',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `bearer ${window.localStorage.getItem('JWT')}`,
        },
        body: JSON.stringify({
          userID: userID,
        }),
      });

      if (response.status === 201) {
        setRequestSent(true);
        setAddButtonText('Request Sent');
        flip();
      }

      if (response.status === 400 || response.status === 404) {
        console.log(response.message);
        setRequestSent(false);
        setAddButtonText('Request Failed');
        flip();
      }
    } catch (err) {
      console.log(err);
    }
  };

  return (
    <div className="add-friend">
      {showForm ? (
        <>
          <form>
            <button onClick={(event) => handleSubmit(event)} type="submit">
              Confirm
            </button>
          </form>
          <button
            onClick={(event) => {
              flip(event);
            }}
          >
            Cancel
          </button>
        </>
      ) : (
        <button>Add Friend</button>
      )}
    </div>
  );
}
