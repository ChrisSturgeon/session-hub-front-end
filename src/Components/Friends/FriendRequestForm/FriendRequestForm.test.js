import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import FriendRequestForm from './FriendRequestForm';
import userEvent from '@testing-library/user-event';

describe('Friend Request Form', () => {
  it('Displays activation button on render', () => {
    render(<FriendRequestForm />);
    expect(screen.getByRole('button')).toHaveTextContent('Add Friend');
  });

  it('matches snapshot', () => {
    const { container } = render(<FriendRequestForm />);
    expect(container).toMatchSnapshot();
  });

  it('Displays form on button activation click', () => {
    render(<FriendRequestForm />);
    const activationButton = screen.getByRole('button', { name: 'Add Friend' });
    userEvent.click(activationButton);
    expect(screen.getByRole('form')).toBeInTheDocument();
  });

  it('Hides form on cancel button click', () => {
    render(<FriendRequestForm />);
    const activationButton = screen.getByRole('button', { name: 'Add Friend' });
    userEvent.click(activationButton);
    const cancelButton = screen.getByRole('button', { name: 'Cancel' });
    userEvent.click(cancelButton);
    expect(screen.queryByRole('form')).not.toBeInTheDocument();
  });
});
