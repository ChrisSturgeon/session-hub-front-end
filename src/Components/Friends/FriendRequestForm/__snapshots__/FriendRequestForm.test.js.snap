// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Friend Request Form matches snapshot 1`] = `
<div>
  <button
    class="add-btn"
  >
    Add Friend
  </button>
</div>
`;
